### PR TITLE
Add mstch methods for namespace scope and class scope

### PR DESCRIPTION
**Pull request by jslee02**
`pybind11` needs to know if the scope is namespace or not. It should uses `.def_submodule()` and `.attr()` for namespace and non-namespace scope, respectively, while `Boost.Python` can use `.attr()` for both cases.

This is just a proof of concept to show how `pybind` template would work with separate mstch objects: namespace scope and non-namespace scope. @psigen Could you guide me how the API should be like?

### CODE REVIEW

**Comment by jslee02** [comment]
@psigen This blocks making progress on `dartpy`. Could you take a look this PR when you have time?

This is basically adding two methods to the clang wrapper: `namespace_` and `scopeWithoutNamespace`. We already have `scope` that returns a mstch node contains all the scope including namespaces and struct/class. However, pybind11 needs to distinguish namespace scope and struct/class scope. The two functions are for it. I know the functions are redundant, but it seems having them would be convenient to support both of Boost.Python and pybind11.

**Comment by psigen** [comment]
Can you give me an example of what the different values the three functions `scope()`, `namespace()` and `scope_without_namespace()` would provide for a few example functions?  I am trying to understand what accessors we would need to generalize this to other binding templates.

**Comment by jslee02** [comment]
```cpp
namespace dart { namespace common {

struct Uri {

enum Type {
AUTHORITY=0,
PATH=1
};

std::string path;
};

} }
```

For example, the three kinds of `PATH` are:
* `scope()`: dart.common.Uri
* `namespace()`: dart.common
* `scope_without_namespace()`: Uri

In pybind11, dart and common should be defined as submodule, and Uri should be attribute. For Boost.Python, they are all attributes.

**Comment by jslee02** [comment]
Without this change, the current pybind11 template will generate following binding for example:
```cpp
void _ZN4dart8dynamics9MeshShape9ColorModeE(pybind11::module& m)
{
    auto sm = m.def_submodule("dynamics").def_submodule("MeshShape");

    ::pybind11::enum_<dart::dynamics::MeshShape::ColorMode>(sm, "ColorMode")
        .value("MATERIAL_COLOR", dart::dynamics::MeshShape::ColorMode::MATERIAL_COLOR)
        .value("COLOR_INDEX", dart::dynamics::MeshShape::ColorMode::COLOR_INDEX)
        .value("SHAPE_COLOR", dart::dynamics::MeshShape::ColorMode::SHAPE_COLOR)
        .export_values();
}
```

Since `MeshShape` is defined as a submodule for an enum inside `MeshShape`, we cannot use or create `MeshShape` class.

File: include/chimera/mstch.h [hunk 1]
```diff
@@ -89,6 +91,18 @@ static_assert(std::is_base_of<clang::NamedDecl, T>::value,
         return chimera::util::constructMangledName(decl_);
     }
 
+    virtual ::mstch::node namespace_()
```
[the remark this answers is not in the document]
Comment (jslee02) [reply include/chimera/mstch.h]: Exactly

**Comment by jslee02** [comment]
I definitely agree with that we need better names. According to [cppreference.com](https://en.cppreference.com/w/cpp/language/scope), relevant scopes for us is just namespace scope and class scope:
* Block scope: this doesn't have a name
* Function parameter scope: not used in our cases
* Function scope: not used in our cases
* Enumeration scope: we use this, but it would be always the last scope so it isn't necessary
* Template parameter scope: not used in our cases
* Point of declaration: not used in our cases

Let me know if I'm missing something here.

**Comment by mxgrey** [comment]
I believe namespace scopes and class scopes are the only scopes where the members can have external linkage. So a concatenation of namespace scope and class scope should cover all use cases for binding C++ symbols (I haven't looked into this PR, but since this is the chimera repo, I assume generating Python bindings for C++ symbols is the goal here).

One thing to make sure of is that you're able to handle nested classes. But there should still be a clean divide between the beginning of the scoping (namespaces) and the end of the scoping (classes), since you can't open a new namespace within a class scope.

In short `namespaceScope + classScope` seems reasonable to me, and I can't think of a better or more descriptive naming.

**Comment by jslee02** [comment]
A simple nested class test is added. Let me add enum class tests in a separate PR with some template updates.

**Comment by jslee02** [comment]
@psigen Function names are updated. Please take a look at the changes when you have a chance.
